File: sdks/python/src/echo_agent_sdk/provider_capabilities_values.py
from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class ProviderCapabilities:
    streaming_tool_calls: bool
    named_sse_events: bool
    reasoning_content: bool
    image_input: bool
    system_as_top_level: bool
    ndjson_streaming: bool
    tool_support: bool
    structured_output: bool
    requires_version_header: bool
    supports_parallel_tool_calls: bool
    supports_tool_choice_none: bool
    tokenizer_name: str | None = None

    @classmethod
    def openai_compatible(cls) -> ProviderCapabilities:
        return cls(True, False, True, True, False, False, True, True, False, True, True)

    @classmethod
    def anthropic(cls) -> ProviderCapabilities:
        return cls(
            False,
            True,
            False,
            True,
            True,
            False,
            True,
            False,
            True,
            True,
            False,
            "claude",
        )

    @classmethod
    def ollama(cls) -> ProviderCapabilities:
        return cls(
            False, False, False, False, False, True, True, False, False, False, False
        )

    @classmethod
    def from_provider_name(cls, name: str) -> ProviderCapabilities:
        normalized = name.lower()
        if normalized == "anthropic":
            return cls.anthropic()
        if normalized == "ollama":
            return cls.ollama()
        return cls.openai_compatible()

File: sdks/typescript/src/provider_capabilities_values.ts
export type ProviderCapabilities = Readonly<{
  streamingToolCalls: boolean;
  namedSseEvents: boolean;
  reasoningContent: boolean;
  imageInput: boolean;
  systemAsTopLevel: boolean;
  ndjsonStreaming: boolean;
  toolSupport: boolean;
  structuredOutput: boolean;
  requiresVersionHeader: boolean;
  supportsParallelToolCalls: boolean;
  supportsToolChoiceNone: boolean;
  tokenizerName?: string;
}>;

function openaiCompatible(): ProviderCapabilities {
  return Object.freeze({ streamingToolCalls: true, namedSseEvents: false, reasoningContent: true, imageInput: true, systemAsTopLevel: false, ndjsonStreaming: false, toolSupport: true, structuredOutput: true, requiresVersionHeader: false, supportsParallelToolCalls: true, supportsToolChoiceNone: true });
}

function anthropic(): ProviderCapabilities {
  return Object.freeze({ streamingToolCalls: false, namedSseEvents: true, reasoningContent: false, imageInput: true, systemAsTopLevel: true, ndjsonStreaming: false, toolSupport: true, structuredOutput: false, requiresVersionHeader: true, supportsParallelToolCalls: true, supportsToolChoiceNone: false, tokenizerName: "claude" });
}

function ollama(): ProviderCapabilities {
  return Object.freeze({ streamingToolCalls: false, namedSseEvents: false, reasoningContent: false, imageInput: false, systemAsTopLevel: false, ndjsonStreaming: true, toolSupport: true, structuredOutput: false, requiresVersionHeader: false, supportsParallelToolCalls: false, supportsToolChoiceNone: false });
}

export const ProviderCapabilities = Object.freeze({
  openaiCompatible,
  anthropic,
  ollama,
  fromProviderName(name: string): ProviderCapabilities {
    switch (name.toLowerCase()) {
      case "anthropic": return anthropic();
      case "ollama": return ollama();
      default: return openaiCompatible();
    }
  },
});
